kernel/src/fs/cgroupfs/systree_node.rs

```diff
@@ -107,15 +259,125 @@ impl CgroupNode {
             SysPerms::DEFAULT_RW_ATTR_PERMS,
         );
         builder.add(SysStr::from("cpu.stat"), SysPerms::DEFAULT_RO_ATTR_PERMS);
+        builder.add(
+            SysStr::from("cgroup.events"),
+            SysPerms::DEFAULT_RO_ATTR_PERMS,
+        );
 
         let attrs = builder.build().expect("Failed to build attribute set");
         Arc::new_cyclic(|weak_self| {
             let fields = BranchNodeFields::new(name, attrs, weak_self.clone());
-            CgroupNode { fields }
+            CgroupNode {
+                fields,
+                inner: RwMutex::new(Some(Inner::default())),
+                depth,
+                populated_count: AtomicUsize::new(0),
+            }
         })
     }
 }
 
+// For process management
+impl CgroupNode {
+    fn propagate_add_populated(&self) {
+        if self.depth <= 1 {
+            return;
+        }
+
+        let mut current_parent = Arc::downcast::<CgroupNode>(self.parent().unwrap()).unwrap();
+        loop {
+            let old_count = current_parent
+                .populated_count
+                .fetch_add(1, Ordering::Relaxed);
+            if old_count > 0 {
+                break;
+            }
+
+            if current_parent.depth == 1 {
+                break;
+            }
+
+            current_parent = Arc::downcast::<CgroupNode>(current_parent.parent().unwrap()).unwrap();
+        }
+    }
+
+    fn propagate_sub_populated(&self) {
+        if self.depth <= 1 {
+            return;
+        }
+
+        let mut current_parent = Arc::downcast::<CgroupNode>(self.parent().unwrap()).unwrap();
+        loop {
+            let old_count = current_parent
+                .populated_count
+                .fetch_sub(1, Ordering::Relaxed);
+            if old_count != 1 {
+                break;
+            }
+
+            if current_parent.depth == 1 {
+                break;
+            }
+
+            current_parent = Arc::downcast::<CgroupNode>(current_parent.parent().unwrap()).unwrap();
+        }
+    }
+
+    /// Performs a read-only operation on the inner data.
+    ///
+    /// If the cgroup node is dead, returns `None`.
+    fn with_inner<F, R>(&self, op: F) -> Option<R>
+    where
+        F: FnOnce(&BTreeMap<Pid, Weak<Process>>) -> R,
+    {
+        let inner = self.inner.read();
+        let inner_ref = inner.as_ref()?;
+
+        Some(op(&inner_ref.processes))
+    }
+
+    /// Performs a mutable operation on the inner data.
+    ///
+    /// If the cgroup node is dead, returns `None`.
+    fn with_inner_mut<F, R>(&self, op: F) -> Option<R>
+    where
+        F: FnOnce(&mut BTreeMap<Pid, Weak<Process>>) -> R,
+    {
+        let mut inner = self.inner.write();
+        let inner_ref = inner.as_mut()?;
+
+        Some(op(&mut inner_ref.processes))
```

Comment:
Won't your subsequent PRs add more fields to the inner object, making the current form inflexible? I'm okay either way. It's up to you whether you revise it in this PR or not.